File: python-exploratory/spiral.py
# spiral.py -- this spirals in a space-filling way

import numpy as np

class Cross(): # cross-scetion
  def __init__(self, arr):
    self.arr = np.array(arr)
    self.shape = np.shape(self.arr)
    self.area_pts = []
    self.area = 0
    self.surface = 0
    self.live_pts = []
    self.dead_pts = []
    self.surface_points = []
    self.pos = self.set_start()
    self.x, self.y = None, None
    self.prev = None
    self.target = self.arr[self.pos[0], self.pos[1]]
    self.no_change = {'prev_dead_pts': None, 'prev_live_pts': None,
                      'count': 0}
    self.spiral_control()

    # here do stuffs
  
  
  def set_start(self):
    adjx = 1 if self.shape[0] % 2 == 0 else 0
    adjy = 1 if self.shape[1] % 2 == 0 else 0
    return [int(i) for i in [self.shape[0]/2-adjx, self.shape[1]/2-adjy]]
  
  
  def no_changes(self):
    
    # assess whether spiral process is done
    if self.no_change['prev_dead_pts'] is None:  # set prev dead pts
      self.no_change['prev_dead_pts'] = len(self.dead_pts)
    elif self.no_change['prev_dead_pts'] == len(self.dead_pts): # increment count
      self.no_change['count'] = self.no_change['count'] + 0.5
      print('Incremented delta to %.1f' %self.no_change['count'])
    else:
      self.no_change['prev_dead_pts'] == len(self.dead_pts) # set new prev dead pts
    
    if self.no_change['prev_live_pts'] is None:  # set prev live pts
      self.no_change['prev_live_pts'] = len(self.live_pts)
    elif self.no_change['prev_dead_pts'] == len(self.live_pts): # increment count
      self.no_change['count'] += 0.5
      print('Incremented delta to %.1f' %self.no_change['count'])
    else:
      self.no_change['prev_live_pts'] = len(self.live_pts) # set new prev live pts
    print(self.no_change)
    return self.no_change['count']
        
  
  
  # directional functions
  move_up = lambda x,y: [x,y-1]
  move_down = lambda x,y: [x,y+1]
  move_right = lambda x,y: [x+1,y]
  move_left = lambda x,y: [x-1,y]
  
  
  
  def check_next(self):
    check = {'right': move_down,      #    -------->
             'left': move_up,         #    ^   x-->
             'up': move_right,        #    |       |
             'down': move_left}       #    <-------  
    next_pt = check[self.prev](self.x, self.y) # get next pt
    if next_pt not in self.dead_pts and next_pt not in self.live_pts:
      if next_pt[0] in range(0,self.shape[0]) and next_pt[1] in range(0,self.shape[1]):
        print(check[self.prev](self.x,self.y))
        print('Next move OK')
        print(self.live_pts)
        return [True, check[self.prev](self.x,self.y)]
      else:
        return [False]
    else:
      print('Next move not OK')
      return [False]
  
    
  
  
  def default_dead_pts(self):
    # make non-target points dead
    for i in range(self.shape[0]):
      for j in range(self.shape[1]):
        if self.arr[i,j] != self.target:
          self.dead_pts.append([i,j]) 
    print('Starting with %i live points and %i dead points' 
          %(len(self.live_pts),len(self.dead_pts)))
    return self
  
  
  def check_live_pt(self, i,j):
    # determine if point should be dead or alive
    dead = 0
    for fn in [move_down, move_up, move_right, move_left]:
      if fn(i,j) in self.dead_pts or fn(i,j) in self.live_pts:
        dead = dead+1 
    if dead == 4 and [i,j] not in self.dead_pts:
      self.dead_pts.append([i,j]) # add it to dead points
      try:
        ind = self.live_pts.index([i,j])
        self.area_pts.append(self.live_pts.pop(ind))
        
      except:
        print('[%i,%i] not in self.live_pts' %(i,j))
      print('Live points: %i, dead points: %i' %(len(self.live_pts),
                                                 len(self.dead_pts)))
      return False
    elif dead == 4 and [i,j] in self.dead_pts:
      return False
    else:
      return True
        
    
  
  def log_pt(self): 
    # log and check if point is dead
    print('Trying to add point %i %i' %(self.x, self.y))
    if self.arr[self.x,self.y] == self.target: # double check
      self.area = self.area + 1
      if self.check_live_pt(self.x, self.y): # if point is alive, add it
        if [self.x,self.y] not in self.live_pts:
          self.live_pts.append([self.x,self.y])
          print('Added point %i %i. Live points: %i' %(self.x, self.y,
                                                     len(self.live_pts)))
      # else it is automatically added to dead_pts
  
  
  def retire_pts(self):
    # if a point is surrounded on all sides by active or dead points
    # (or both), make it dead
    # print(self.shape)
    for i in range(self.shape[0]):
      for j in range(self.shape[1]):
        if not self.check_live_pt(i,j):
          print('Live points: %i, dead points: %i' %(len(self.live_pts),
                                                     len(self.dead_pts)))
    
        
  
  
  def spiral(self):
    # continue to spiral until an edge is hit; then try other directions
    # when all directions hit an edge, move to another point
    from itertools import cycle
    self.area = self.area + 1
    self.x, self.y = self.pos[0], self.pos[1]
    self.prev = 'up' # set fake previous move
    self.log_pt()
    moves = [move_down, move_left, move_up, move_right]
    prevs = ['right','down','left','up']
    _moves = cycle(moves)
    _prevs = cycle(prevs)
    delta = self.no_changes()
    # cycle through points
    while len(self.live_pts) > 2 or delta < 20:
      count = 0
      self.retire_pts()
      delta = self.no_changes()
      print('Delta is %.1f' %delta)
      # take a point until exhausted
      self.x, self.y = self.live_pts[0]
      print('Starting new point %i %i' %(self.x, self.y))
      while count < 4: # if next exists, go to it
        if self.check_next()[0]:
          # print(self.check_next())
          try:
            self.x, self.y = self.check_next()[1][0], self.check_next()[1][1]
          except:
            self.x, self.y = self.check_next()[1][0][0], self.check_next()[1][0][1]
          self.log_pt()
          count = 0 # reset count, found a new point
          
          next_move = next(_moves)
          self.prev = next(_prevs)
        else:
          next_move = next(_moves)
          self.prev = next(_prevs)
          count = count + 1
          if count ==4:
            print(count)
          # otherwise, check the others and increase count
  
  
  def spiral_control(self):
    # control the class
    self.default_dead_pts()
    self.spiral()
    print(self.dead_pts)
    return self
        
        
        

File: morphology/cpp/image/Spiral.cpp
// a simple test of stuff

#include <fstream>
#include <iostream>
#include <vector>
#include <cassert>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <time.h>
#include <boost/tuple/tuple.hpp>
#include <boost/multi_array.hpp>
#include <eigen3/Eigen/Dense>
// #include "imageMatrix_class.h" // Not needed?
#include "Spiral.h" // This is just constructors for the larger class Spiral,
                    // not all Spiral methods have been added to this yet.


using boost::tuple;
using namespace Eigen;

typedef std::vector< boost::tuple<int, int> > tuple_list;



////////////////////////////////////////////////////////////////////////
// testing



// Constructor
Spiral::Spiral(MatrixXi array, std::vector<int> ddims, boost::tuple<int, int> spt) {
  arr = array;
  for (auto i = ddims.begin(); i != ddims.end(); ++i) {
    int temp = *i;
    dims.push_back(temp);
  }
  int t_x, t_y;
  t_x = boost::get<0>(spt);
  t_y = boost::get<1>(spt);
  startpt = {t_x, t_y};
  x = t_x;
  y = t_y;
  std::cout << "Uninitialized target was: " << tgt << ", new target is: ";
  tgt = arr(t_x, t_y); // Target should be binary, ==1
  std::cout << tgt << std::endl;
  tuple_list dead_pts;
  tuple_list live_pts;
  int delta_dead = 0;
  int delta_live = 0;
  int delta_count = 0;
  int tolerance; // If delta_count > tolerance, spiral exits
  prev = 0; // 0=right->down, 1=left->up, 2=up->right, 3=down->left
  int x, y;
  std::vector<int> prev_pt;
  int area = 0;
  std::cout << arr << std::endl;
  std::cout << "(" << boost::get<0>(startpt) << "," <<
    boost::get<1>(startpt) << ")\n" << std::endl;
  return;
}



// Assess whether spiral process is done
void Spiral::no_changes() {
  // Set current length of live and dead pts
  if (delta_dead == 0) {
    delta_dead = dead_pts.size();
  }
  if (delta_live == 0) {
    delta_live = live_pts.size();
  }
  // Check live and dead pts; if they're the same, increment the count
  // Otherwise set the new delta level
  if (dead_pts.size() == delta_dead) {
    delta_count = delta_count + 1;
  }
  else {
    delta_dead = dead_pts.size();
  }
  if (live_pts.size() == delta_live) {
    delta_count = delta_count + 1;
  }
  else {
    delta_live = live_pts.size();
  }
}



bool is_element(tuple_list list, boost::tuple<int, int> tup) {
  std::vector<int> test_vec = {boost::get<0>(tup), boost::get<1>(tup)};
  // Pull out the elements of the test tuple (operator==)
  int tup_0 = boost::get<0>(tup);
  int tup_1 = boost::get<1>(tup);
  for (tuple_list::const_iterator i = list.begin(); i != list.end(); ++i) {
    // Pull out the element of the possible tuple
    int elem_0 = i->get<0>();
    int elem_1 = i->get<1>();
    if (tup_0 == elem_0 && tup_1 == elem_1) {
      return true;
    }
  }
  return false;
}



void show_tup(boost::tuple<int, int> tup) {
  bool newline = false;
  std::cout << "(" << tup.get<0>() << "," << tup.get<1>() <<")";
  if (newline) {std::cout << "\n";}
}



// Show some elements of a tuple_list
void show_tup_list(tuple_list tlist) {
  std::cout << "Summary of "<< tlist.size() << " elements: \n";
  for (tuple_list::const_iterator i = tlist.begin(); i != tlist.begin()+5; ++i) {
    boost::tuple<int, int> temp_tup (i->get<0>(), i->get<1>());
    show_tup(temp_tup);
    std::cout << "\n";
  }
  std::cout << " ... \n";
  for (tuple_list::const_iterator i = tlist.end()-5; i != tlist.end(); ++i) {
    boost::tuple<int, int> temp_tup (i->get<0>(), i->get<1>());
    show_tup(temp_tup);
    std::cout << "\n";
  }
}



void show_vec(std::vector<int> vec) {
  std::cout << "(";
  for (std::vector<int>::const_iterator i = vec.begin(); i != vec.end(); ++i) {
    std::cout << *i << " ";
  }
  std::cout << ")\n";
}



// Make all non-target points dead *unless* that means making >95% live
// (then switch) // This function is currently being tested in main()
void Spiral::default_dead_pts() {
  int dead_count = 0;
  int other;
  // Iterate through the 2-D array
  for (int i = 0; i < dims[0]; i++) { 
    for (int j = 0; j < dims[1]; j++) {
      if (arr(i,j) != tgt) {
        other = arr(i,j);
        // show_tup(boost::tuple<int, int> (i,j));
        dead_pts.push_back(boost::tuple<int, int> (i,j));
        dead_count = dead_count + 1;
      }
    }
  }
  
  if (dead_count < (dims[0]*dims[1])*0.15) {
    for (int i = 0; i < dead_pts.size(); i++) {
      dead_pts.pop_back();
    }
    dead_count = 0;
    for (int i = 0; i < dims[0]; i++) {
      for (int j = 0; j < dims[1]; j++) {
        if (arr(i,j) == tgt) {
          dead_pts.push_back( boost::tuple<int, int>(i,j) );
          dead_count = dead_count + 1;
          // Switch target and other
          tgt = other;
        }
      }
    }
  } 
  
  std::cout << "number of dead points: " << dead_pts.size() << std::endl;
  // iterate through dead_pts and show them
  show_tup_list(dead_pts);
}



// Check to see if the next point exists (in the array) and is not
// already in live_pts or dead_pts
boost::tuple<bool, int, int> Spiral::check_next() {
  std::vector<int> next_pt;
  // std::cout << "previous pt: [" << prev_pt[0] <<","<< prev_pt[1] <<"]\n";
  switch (prev) {
    // 0 : prev=right, next=down
    case 0 : next_pt = move_down(prev_pt);
             // prev = 3; // Don't actually advance the next
             break;
    // 1 : prev = left, next=up
    case 1 : next_pt = move_up(prev_pt);
             // prev = 2;
             break;
    // 2 : prev=up, next=right
    case 2 : next_pt = move_right(prev_pt);
             // prev = 0;
             break;
    // 3 : prev=down, next=left
    case 3 : next_pt = move_left(prev_pt);
             // prev = 1;
             break;
  }
  // If point not in live or dead points ...
  boost::tuple <int, int> next_pt_t (next_pt[0], next_pt[1]);
  if (! is_element(dead_pts, next_pt_t)) {
    if (! is_element(live_pts, next_pt_t)) {
      /* find(v.begin(), v.end(), x) == v.end() is true if x NOT IN v
         find(v.begin(), v.end(), x) != v.end() is true if x IN v */
      // ... and point is within the bounds of the array ...
      if (next_pt[0] > 0 && next_pt[0] < dims[0]) {
        if (next_pt[1] > 0 && next_pt[1] < dims[1]) {
          // ... then return that the point is okay.
          return boost::tuple<bool, int, int> (true, boost::get<0>(next_pt_t),
                                    boost::get<1>(next_pt_t)); // Return as std::vector or tuple?
        }
      }
    }
  }
  // Else, return false and some bullshit
  return boost::tuple<bool, int, int> (false, 0,0);
}



/* This is called by log_pt, which tries to add a live_pt if it should
   be added (not already in live_pts or dead_pts). If a false is
   returned, that means the point is already in one of those lists.
   A true returned means add it to live (but there was insufficient info
   for check_live_pt to add it). */
bool Spiral::check_live_pt(std::vector<int> potpt) {
  // If it's off-target, can return it as false immediately
  auto potpt_t = boost::make_tuple(potpt[0], potpt[1]);
  if (arr(potpt[0], potpt[1]) != tgt) {
    // Add it to dead points if it's not already there
    if (! is_element(dead_pts, potpt_t)) {
      dead_pts.push_back(potpt_t);
    }
    return false;
  }
  // Else, check if it's surrounded on any side by live pts
  int live = 0;
  std::vector<int> next_pt1, next_pt2, next_pt3, next_pt4;
  next_pt1 = move_up(potpt);                                     // Up
  auto next_pt_t1 = boost::make_tuple(next_pt1[0], next_pt1[1]);
  if (is_element(live_pts, next_pt_t1)) { live = live + 1;}
  next_pt2 = move_down(potpt);                                   // Down
  auto next_pt_t2 = boost::make_tuple(next_pt2[0], next_pt2[1]);
  if (is_element(live_pts, next_pt_t2)) { live = live + 1;}
  next_pt3 = move_left(potpt);                                   // Left
  auto next_pt_t3 = boost::make_tuple(next_pt3[0], next_pt3[1]);
  if (is_element(live_pts, next_pt_t3)) { live = live + 1;}
  next_pt4 = move_right(potpt);                                  // Right
  auto next_pt_t4 = boost::make_tuple(next_pt4[0], next_pt4[1]);
  if (is_element(live_pts, next_pt_t4)) { live = live + 1;}
  // If at least one side is live and the point itself isn't dead, make live
  // This means the point has already been added, so return false
  if (live >= 1 && ! is_element(dead_pts, potpt_t)) {
    // As long as it's not already in live_pts
    if (! is_element(live_pts, potpt_t)) {
      live_pts.push_back(potpt_t);
      // And increment the area?
    }
    return false;
  }
  return true; // Else, return true (try to add the point)
}



// Log the point as a live_pt
void Spiral::log_pt() {
  // Check current point
  if (arr(x,y) == tgt) {
    // std::cout << "good point found: (" << x << "," << y << ")\n";
    area = area + 1;
    std::vector<int> potpt;
    potpt.push_back(x);
    potpt.push_back(y);
    if (check_live_pt(potpt)) {
      auto potpt_t = boost::make_tuple(x, y);
      if (! is_element(live_pts, potpt_t)) {
        live_pts.push_back(potpt_t);
      }
    }
  }
}



////////////////////////////////////////////////////////////////////////
// Direction in-line functions

// Move up
inline std::vector<int> move_up(std::vector<int> prev_pt) {
  std::vector<int> next_pt = {prev_pt[0], prev_pt[1]-1};
  return next_pt;
}
// Move down
inline std::vector<int> move_down(std::vector<int> prev_pt) {
  std::vector<int> next_pt = {prev_pt[0], prev_pt[1]+1};
  return next_pt;
}
// Move right
inline std::vector<int> move_right(std::vector<int> prev_pt) {
  std::vector<int> next_pt = {prev_pt[0]+1, prev_pt[1]};
  return next_pt;
}
// Move left
inline std::vector<int> move_left(std::vector<int> prev_pt) {
  std::vector<int> next_pt = {prev_pt[0]-1, prev_pt[1]};
  return next_pt;
}



////////////////////////////////////////////////////////////////////////
// Spiral
//      ------->
//     ^   x-->
//     |       |
//      <------
//
// Helper function -- cycle between 0 -(down)-> 3 -(left)-> 1 -(up)-> 2 -(right)-> 0
int cycle(int prev) {
  if (prev > -1 && prev < 4) {
    switch(prev) {
      case 0 : return 3;
      case 3 : return 1;
      case 1 : return 2;
      case 2 : return 0;
    }
  }
  return -1;
}



// Spiral
void Spiral::spiral() {
  srand (time(NULL)); // Seed random number
  tolerance = 2000; // Set some tolerance
  // Add the first area
  area = area + 1;
  // int r_x, r_y; // Random numbers for start points
  x = startpt.get<0>();
  y = startpt.get<1>();
  prev = 2; // Set 'fake' previous move to 'up'
  log_pt(); // Log the start point
  //no_changes(); // Log the state of the lists
  while (live_pts.size() > 0) {
    int count = 0;
    no_changes();
    // Choose a random point
    x = rand() % dims[0];
    y = rand() % dims[1];
    // std::cout << "rand point: (" << x << "," << y << ")\n";
    log_pt();
    prev_pt = {x, y};
    // Check each direction for the new point
    while (count < 4) { 
      auto next_t = check_next();
      // If the next point exists
      if (boost::get<0>(next_t)) {
        x = boost::get<1>(next_t); // Set the new x and y values
        y = boost::get<2>(next_t); 
        log_pt();
        count = 0; // Reset the count -- found a new point
        // Advance the cycles
        prev = cycle(prev);
      }
      // If the next point is not valid, cycle and increment
      else {
        prev = cycle(prev);
        count = count + 1;
      }
      // Check tolerance; if it exceeds then quit, otherwise continue
      if (delta_count > tolerance) return;
    }
  }
}

